File: UnityMcpBridge/Editor/Tools/CaptureView.cs
using System;
using Newtonsoft.Json.Linq;
using UnityEditor;
using UnityEngine;
using UnityMcpBridge.Editor.Helpers;

namespace UnityMcpBridge.Editor.Tools
{
    /// <summary>
    /// Captures screenshots of the Game or Scene view and returns them as base64 strings.
    /// </summary>
    public static class CaptureView
    {
        /// <summary>
        /// Main handler for view capture actions.
        /// </summary>
        public static object HandleCommand(JObject @params)
        {
            string action = @params["action"]?.ToString().ToLower() ?? "capture";
            string target = @params["target"]?.ToString()?.ToLower() ?? "game";

            switch (action)
            {
                case "capture":
                    return Capture(target);
                default:
                    return Response.Error($"Unknown action: '{action}'. Valid action is 'capture'.");
            }
        }

        private static object Capture(string target)
        {
            try
            {
                Texture2D screenshot = null;

                if (target == "scene")
                {
                    SceneView sceneView = SceneView.lastActiveSceneView;
                    if (sceneView == null)
                    {
                        return Response.Error("No active Scene view found.");
                    }

                    Camera cam = sceneView.camera;
                    int width = Mathf.RoundToInt(sceneView.position.width);
                    int height = Mathf.RoundToInt(sceneView.position.height);
                    RenderTexture rt = new RenderTexture(width, height, 24);
                    cam.targetTexture = rt;
                    cam.Render();
                    RenderTexture.active = rt;
                    screenshot = new Texture2D(width, height, TextureFormat.RGB24, false);
                    screenshot.ReadPixels(new Rect(0, 0, width, height), 0, 0);
                    screenshot.Apply();
                    cam.targetTexture = null;
                    RenderTexture.active = null;
                    UnityEngine.Object.DestroyImmediate(rt);
                }
                else // default to game view
                {
                    screenshot = ScreenCapture.CaptureScreenshotAsTexture();
                }

                if (screenshot == null)
                {
                    return Response.Error("Failed to capture screenshot.");
                }

                byte[] pngData = screenshot.EncodeToPNG();
                string base64 = Convert.ToBase64String(pngData);
                int widthOut = screenshot.width;
                int heightOut = screenshot.height;
                UnityEngine.Object.DestroyImmediate(screenshot);

                return Response.Success(
                    "Captured screenshot.",
                    new { base64 = base64, width = widthOut, height = heightOut }
                );
            }
            catch (Exception e)
            {
                return Response.Error($"Capture failed: {e.Message}");
            }
        }
    }
}

File: UnityMcpServer/src/tools/capture_view.py
from mcp.server.fastmcp import FastMCP, Context
from typing import Dict, Any
from unity_connection import get_unity_connection


def register_capture_view_tools(mcp: FastMCP):
    """Register the capture_view tool with the MCP server."""

    @mcp.tool()
    def capture_view(ctx: Context, target: str = "game") -> Dict[str, Any]:
        """Capture a screenshot of the specified Unity view.

        Args:
            ctx: The MCP context.
            target: View to capture ('game' or 'scene').

        Returns:
            Dictionary with success status, message and data containing the
            base64 encoded PNG image.
        """
        params = {"action": "capture", "target": target}
        try:
            response = get_unity_connection().send_command("capture_view", params)
            if response.get("success"):
                return {
                    "success": True,
                    "message": response.get("message", "Captured view."),
                    "data": response.get("data"),
                }
            else:
                return {
                    "success": False,
                    "message": response.get("error", "Unknown error capturing view."),
                }
        except Exception as e:
            return {"success": False, "message": f"Python error capturing view: {str(e)}"}
